File: 86/pe86.cpp
#include <iostream>
#include <cmath>

int IsSquare(const int n) {
  const auto sqrt = std::sqrt(n);
  return sqrt == std::floor(sqrt);
}

int FindM(const int solutions) {
  auto count = 0;
  auto M = 0;
  while (true) {
    for (auto w = 1; w <= M; ++w) {
      for (auto h = w; h <= M; ++h) {
        if (IsSquare(std::pow(w + h, 2) + std::pow(M, 2))) {
          ++count;
        }
      }
    }
    std::cout << "M:" << M << " count:" << count << std::endl;
    if (count > solutions) {
      return M;
    }
    ++M;
  }
}

int main() {
  std::cout << FindM(2000) << std::endl;
  std::cout << FindM(1000000) << std::endl;
}

File: 86/pe86.py
squares = set([n ** 2 for n in xrange(1, 100000)])
print "is_square(n) is ready."

def is_square(n):
    return n in squares

def find_M(solutions):
    count = 0
    M = 1
    while True:
        for w in xrange(1, M + 1):
            for h in xrange(w, M + 1):
                if is_square((w + h) ** 2 + M ** 2):
                    count += 1
        print "M:", M, "count:", count
        if count > solutions:
            return M
        M += 1

print find_M(2000)
print find_M(1000000)
